src/main/java/jenkins/plugins/git/traits/DiscoverOtherRefsTrait.java

```diff
@@ -0,0 +1,154 @@
+import jenkins.plugins.git.GitSCMBuilder;
+import jenkins.plugins.git.GitSCMSource;
+import jenkins.plugins.git.GitSCMSourceContext;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.trait.SCMBuilder;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceTrait;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import jenkins.scm.impl.trait.Discovery;
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jgit.lib.Constants;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import static jenkins.plugins.git.AbstractGitSCMSource.REF_SPEC_REMOTE_NAME_PLACEHOLDER_STR;
+
+public class DiscoverOtherRefsTrait extends SCMSourceTrait {
+
+    private final String ref;
+    private String nameMapping;
+
+    @DataBoundConstructor
+    public DiscoverOtherRefsTrait(String ref) {
+        if (StringUtils.isEmpty(ref)) {
+            throw new IllegalArgumentException("ref can not be empty");
+        }
+        this.ref = StringUtils.removeStart(StringUtils.removeStart(ref, Constants.R_REFS), "/");
+        setDefaultNameMapping();
+    }
+
+    //for easier testing
+    public DiscoverOtherRefsTrait(String ref, String nameMapping) {
+        this(ref);
+        setNameMapping(nameMapping);
+    }
+
+    public String getRef() {
+        return ref;
+    }
+
+    String getFullRefSpec() {
+        return new StringBuilder("+")
+        .append(Constants.R_REFS).append(ref)
+        .append(':').append(Constants.R_REMOTES)
+        .append(REF_SPEC_REMOTE_NAME_PLACEHOLDER_STR)
+        .append('/').append(ref).toString();
+    }
+
+    public String getNameMapping() {
+        return nameMapping;
+    }
+
+    @DataBoundSetter
+    public void setNameMapping(String nameMapping) {
+        if (StringUtils.isEmpty(nameMapping)) {
+            setDefaultNameMapping();
+        } else {
+            this.nameMapping = nameMapping;
+        }
+    }
+
+    private void setDefaultNameMapping() {
+        this.nameMapping = null;
+        String[] paths = ref.split("/");
+        for (int i = 0; i < paths.length; i++) {
+            if("*".equals(paths[i]) && i > 0) {
+                this.nameMapping = paths[i-1] + "-@{1}";
+                break;
+            }
+        }
+        if (StringUtils.isEmpty(this.nameMapping)) {
+            if (ref.contains("*")) {
+                this.nameMapping = "other-@{1}";
+            } else {
+                this.nameMapping = "other-ref";
+            }
+        }
+    }
+
+    @Override
+    protected void decorateContext(SCMSourceContext<?, ?> context) {
+        GitSCMSourceContext c = (GitSCMSourceContext) context;
+        c.withRefSpec(getFullRefSpec());
+        c.wantOtherRef(new GitSCMSourceContext.WantedOtherRef(this.ref, this.nameMapping));
+    }
+
+    /**
+     * Our descriptor.
+     */
+    @Extension
+    @Discovery
+    public static class DescriptorImpl extends SCMSourceTraitDescriptor {
```

Comment:
If we remove the `@Extension` and make this class `abstract` then it would be trivial to add this trait using an extension plugin (basically extend the descriptor and add `@Extension`)